### PR TITLE
Implement API endpoint that returns name of collection containing document having a given `id`

### PR DESCRIPTION
**Pull request by eecavanna**
# Description

In this branch, I implemented a new API endpoint. Its behavior is described in its docstring, shown here:

```py
@router.get("/nmdcschema/ids/{doc_id}/collection-name")
def get_collection_name_by_doc_id(
    doc_id: str,
    mdb: MongoDatabase = Depends(get_mongo_db),
):
    r"""
    Returns the name of the collection, if any, containing the document having the specified `id`.

    This endpoint uses the NMDC Schema to determine the schema class of which an instance could have
    the specified value as its `id`; and then uses the NMDC Schema to determine the names of the
    `Database` slots (i.e. Mongo collection names) that could contain instances of that schema class.

    This endpoint then searches those Mongo collections for a document having that `id`.
    If it finds one, it responds with the name of the collection containing the document.
    If it does not find one, it response with an `HTTP 404 Not Found` response.
    """
```

Fixes https://github.com/microbiomedata/nmdc-runtime/issues/531

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration, if it is not simply `make up-test && make test-run`.

- [x] I implemented an automated test targeting the new functionality and that test ran/passed on GitHub Actions:
      image

**Configuration Details**: none

# Checklist:

- [x] My code follows the style guidelines of this project (have you run `black nmdc_runtime/`?)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (in `docs/` and in <https://github.com/microbiomedata/NMDC_documentation/>?)
- [x] I have added tests that prove my fix is effective or that my feature works, incl. considering downstream usage (e.g. <https://github.com/microbiomedata/notebook_hackathons>) if applicable.
- [ ] New and existing unit and functional tests pass locally with my changes (`make up-test && make test-run`)
  > I delegated this test running to GitHub Actions, which passes

### CODE REVIEW

**Comment by eecavanna** [comment]
Note: I wrote this code in my web browser, on `github.dev`. I haven't tested it in **any** way yet.

**Comment by eecavanna** [comment]
@aclum, is this the type of HTTP response you had in mind for this endpoint that we discussed (the endpoint that, for a given `id`, determines the collection—if any—in which a document having that `id` would reside, and determines the schema class—if any—of which an instance would have that as its `id`)?

```json
    {
        "id": "nmdc:sty-10-123456abcdef",
        "collection_name": "study_set",
        "class_name": "Study",
    }
```

CC: @sujaypatil96

**Comment by aclum** [comment]
Yes. I think this would also be useful for the data portal, I have a ticket that has been in for a while to be able to search by an nmdc identifier. https://github.com/microbiomedata/nmdc-server/issues/964 cc @naglepuff @marySalvi @jeffbaumes

**Comment by dwinston** [comment]
What is expected for an `id` that looks like e.g. `nmdc:clsite-99-123abc`? From `/nmdcschema/typecodes`, `clsite` is the typecode for a `CollectingBiosamplesFromSite` object, but this can be in either the `collecting_biosamples_from_site_set` (https://github.com/microbiomedata/nmdc-schema/blob/v10.5.3/nmdc_schema/nmdc_materialized_patterns.schema.json#L3531) or `planned_process_set` (https://github.com/microbiomedata/nmdc-schema/blob/v10.5.3/nmdc_schema/nmdc_materialized_patterns.schema.json#L3653) slots, according to the schema. So in this case, both collection names should be returned, no?

I have updated the PR to reflect this and return a list of collections.

**Comment by aclum** [comment]
The issue of being valid for multiple collections is fixed in berkeley schema. The original feature of this endpoint was to return which collection the ID is in so if it is ambiguous for now it should check which collection the document actually resides in.

File: nmdc_runtime/api/endpoints/nmdcschema.py [hunk 1]
```diff
@@ -131,6 +134,70 @@ def get_by_id(
...
+    match = pattern.search(hypothetical_doc_id)
+    typecode_portion = match.group(1) if match else None
+
+    if typecode_portion is None:
+        return None  # abort
+
+    # Determine the schema class, if any, of which the specified `id` could belong to an instance.
+    schema_class_name = None
+    for typecode in typecodes():
+        if typecode_portion == typecode["name"]:
```
Comment (eecavanna) [inline nmdc_runtime/api/endpoints/nmdcschema.py:164]: Thank you, @dwinston! (Maybe needless to say), this change you made fixes a bug in the code I wrote (oops).

**Comment by eecavanna** [comment]
> The original feature of this endpoint was to return which collection the ID is in so if it is ambiguous for now it should check which collection the document actually resides in.

This endpoint only uses the schema, not the database. The `class_name` it returns is the name of the schema class, if any, for which the specified value would be a valid `id` (on an instance of that class). The `collection_names` **list** it returns is a list of the names of all the "collections" (technically, the names of slots of the `Database` schema class), if any, that could contain a document having that value as its `id`.

In other words, it currently doesn't do this:

> if it is ambiguous for now it should check which collection the document actually resides in.

Instead, it returns a list of all collections that — according to the schema — a document having that `id` could reside in.

CC: @aclum

**Comment by eecavanna** [comment]
Sounds to me like one of the requirements for this endpoint is to also indicate which collection, if any, a document having the specified `id` resides in.

**Comment by eecavanna** [comment]
> Sounds to me like one of the requirements for this endpoint is to also indicate which collection, if any, a document having the specified id resides in.

**Update:** I have added this feature to the endpoint. Example response:

```json
{
    "id": "nmdc:sty-1-foobar",
    "class_name": "Study",
    "collection_names": ["study_set"],
    "containing_collection_name": "study_set"
}
```

`"containing_collection_name": "study_set"` means that a document having the specified `id` was found in that collection in the Mongo database.

**Comment by eecavanna** [comment]
I am ready for this PR branch to be reviewed/merged.

File: nmdc_runtime/api/endpoints/nmdcschema.py [hunk 2]
```diff
@@ -131,6 +134,100 @@ def get_by_id(
...
+                f"but no compatible collection names were found."
+            ),
+        )
+
+    # Use the Mongo database to determine which of those collections a document having that `id` actually
+    # resides in, if any. If multiple collections contain such a document, report only the first one.
+    containing_collection_name = None
+    for collection_name in collection_names:
+        collection = mdb.get_collection(name=collection_name)
+        if collection.count_documents(dict(id=hypothetical_doc_id)) > 0:
```
Comment (eecavanna) [inline nmdc_runtime/api/endpoints/nmdcschema.py:219]: TODO: Pass a `limit` kwarg to the `count_documents` method so it stops counting after encountering the first matching document.

**Review by sujaypatil96 — DISMISSED** [review]
I tested the `/nmdcschema/ids/{hypothetical_doc_id}/class-and-collection-names` endpoint with 3 ids.

* Case 1: An Extraction id which exists in the database (for testing, in my local Mongo), and can (theoretically) be part of multiple collections
Screenshot 2024-06-13 at 10 12 02 AM


* Case 2: A study id which can technically be part of only one collection
Screenshot 2024-06-13 at 10 12 57 AM

* Case 3: Non-conformant id
Screenshot 2024-06-13 at 10 13 27 AM

<hr/>

It worked like a charm in all three cases, so I'm happy with its behavior and am ready for this PR to be merged 🚀

**Comment by eecavanna** [comment]
Hi @aclum, are you OK with how this endpoint behaves? If so (and someone approves it via GitHub's Review mechanism), I'll merge it in. I accidentally invalidated @sujaypatil96's approval by making an additional commit.

@sujaypatil96 added screenshots of some example request/response pairs [above](https://github.com/microbiomedata/nmdc-runtime/pull/532#pullrequestreview-2116490388)—thanks, @sujaypatil96!

**Review by PeopleMakeCulture — APPROVED** [review]
lgtm. thank you for writing tests!

**Comment by PeopleMakeCulture** [comment]
@eecavanna You're good to merge. If/when there's time, I'd like to consolidate and replace these utils with your logic. (non-blocking):
- `nmdc_runtime.api.core.metadata.map_id_to_collection` 
- `nmdc_runtime.util.collection_name_to_class_names`
- `nmdc_runtime.api.db.mongo.nmdc_schema_collection_names`

**Comment by eecavanna** [comment]
Based on @aclum's feedback during today's infrastructure meeting, I will update the endpoint as follows:
- [ ] 1. If a document having the specified `id` exists in any of the collections the schema says it can exist in, return the name of that collection. If it happens to exist in multiple collections, only return the name of one collection that it exists in.
- [ ] 2. If no document having the specified `id` exists in any of the collections the schema says it can exist in, return a "null-ish" response (TBD).

**Comment by eecavanna** [comment]
Sorry about all the thrash here! Getting the automated tests running locally was not happening smoothly for me (I think one of the tests hung last night), so I've been relying on the GHA workflow to run them for me. A downside of that is that the test failures generate notifications—at least to me.

**Comment by eecavanna** [comment]
Hi @aclum, I updated the API response based upon our conversation earlier today.

The new behavior is:

1. **Scenario 1:** User specifies the `id` of `nmdc:sty-1-foobar` to the API. A document having that `id` exists in the `study_set` collection. The API responds with:
   ```json
   {
      "id": "nmdc:sty-1-foobar",
      "collection_name": "study_set",
   }
   ```

2. **Scenario 2:** User specifies the `id` of `nmdc:sty-1-nonex` to the API. A document having that `id` does not exist in any collections (that the schema says it can exist in). The API responds with a `HTTP 404 Not Found` response.

I'm ready for this PR branch to be reviewed/merged in.

**Review by aclum — APPROVED** [review]
